### PR TITLE
Add metadata struct and add it to Constant

### PR DESCRIPTION
**Pull request by Kieranoski702**
# PR Description
This is the start of the `Metadata` functionality. I've created the metadata struct within `metadata.rs` and in order to test if it can work I added it to `Constant` inside the `Expression` enum. This meant another 6 files had to be changed that used `Constant` which took quite a bit of work to change. Even after all this I have still left two TODOs (one in `parse.rs` and another in `rules/base.rs`) as I think I need some way to grab metadata from previous expressions to allow these to work.

While this does compile, the tests are currently failing. The one that fails right now is the XYZ auto-generated integration test. I have tried to fix this but I do not know enough about the integration tests or how they are generated. Would you mind shedding some light on why this might be failing @gskorokhod?

In terms of next steps, all the other parts of the enum have to have `Metadata` added to them. This will take me quite a long time to do on my own I think so if anybody has spare time I have created a quick start guide to implementing metadata on an `Expression` enum variant at the end of this PR description. I can get through everything by myself but it will just take a significant amount of time, I think, so any help is greatly appreciated.

I have tagged @ozgurakgun, @niklasdewally and @lixitrixi as reviewers as I have directly changed your code, if I have made a mistake in your implementations then please let me know!

# Implementing metadata on an `Expression` enum variant mini guide:

1. Add the `Metadata` struct as a parameter of the variant, preferably as the first parameter. `Constant` example:
```rust
Constant(Metadata, Constant)
```

2. Fix the errors everywhere that variant has been used. This will mostly consist of using `Metadata::new()` when a new blank Metadata struct is needed or passing the existing metadata in an expression to a new expression. One example where `Constant` changed:
```rust
let four = ConjureExpression::Constant(Metadata::new(), ConjureConstant::Int(4));
```

3. Ensure functionality is the same. Run `cargo test` and fix the errors within the test files to ensure that the functionality of conjure oxide is the same with your new changes. Request additional help as needed

4. Mark unsupported functionality with a TODO comment. There will be situations where the `Metadata` struct and module don't yet have the features to convert the current functionality fully, mark these areas with TODO comments and I will implement the required features. For example, I need to create some way for the metadata of multiple constants to be "merged" so that creating a sum of integer constants works properly:
```rust
// TODO (kf77): Get existing metadata instead of creating a new one
new_exprs.push(Expr::Constant(Metadata::new(), Const::Int(sum)));
```

### CODE REVIEW

**Comment by gskorokhod** [comment]
I haven't looked at the code yet, but off the top of my head, it can be one of the following:
- Either the AST has changed (e.g. you added a new metadata field to our variants) but the test still expects the old AST structure 
- Or the Minion solver adapter doesn't know what to do with this extra metadata field 

I'll try and run it on my machine today tho and have a proper look!

**Review by lixitrixi — APPROVED** [review]
Looks bueno!

**Comment by niklasdewally** [comment]
> I haven't looked at the code yet, but off the top of my head, it can be one of the following:
> 
> * Either the AST has changed (because you added a new metadata field to our variants) but the test still expects the old AST structure
> * Or the Minion solver adapter doesn't know what to do with this extra metadata field
> 
> I'll try and run it on my machine today tho and have a proper look!

I just rebased this branch off of #217 locally, and it seems to be the ast parse step that is the problem

```
RUN: model parse check
Error: Error("invalid type: map, expected tuple variant Expression::Constant", line: 30, column: 24)
```

**Comment by niklasdewally** [comment]
@Kieranoski702 @gskorokhod Deleting the expected tests and recreating them from the generated fixes them. As im forked off the new tester stuff I wrote,  I can also confirm that this is runnable by Minion too :)

File: conjure_oxide/src/parse.rs [hunk 1]
```diff
@@ -247,12 +248,16 @@ fn parse_vec_op(
...
-            Some(Expression::Constant(Constant::Int(
-                int["ConstantInt"].as_array()?[1]
-                    .as_i64()?
-                    .try_into()
-                    .unwrap(),
-            )))
+            Some(Expression::Constant(
+                Metadata::new(),
+                Constant::Int(
+                    //TODO (kf77): Get metadata from JSON?
```
Comment (ozgurakgun) [inline conjure_oxide/src/parse.rs:254]: I don't think there is anything sensible to extract from the JSON as metadata. I feel we will need a native parser soon (not that soon, maybe next year)

**Review by ozgurakgun — COMMENTED** [review]
Looks good to me. Not even as ugly as we feared (yet?)

**Review by niklasdewally — COMMENTED** [review]
As discussed in person, this is definitely the way to go 👍🏻 

One problem (related to the tests failing) is that we are serialising internal metadata stuff into our JSON. This means that everytime we make some small metadata change, we will break integration tests. I think metadata is internal, so we should not hold ourselves to specific values or syntax for this in the integration tests. We also probably won't have the same metadata at the "rewrite check" stage every time, and this (in my opinion) is expected behaviour. Maybe we could hide this from the JSON output?

The fix to this would be to hide the metadata field from serialisation somehow (or maybe only print it with the presence of a -DEBUG config option?)


Would be curious to know what tester people (@ozgurakgun @gskorokhod) think about this.

**Comment by ozgurakgun** [comment]
> I think metadata is internal, so we should not hold ourselves to specific values or syntax for this in the integration tests. We also probably won't have the same metadata at the "rewrite check" stage every time,

I say keep it for now and update the expected files. Following the "err on the side of testing too much" rule. 

I'll be surprised if metadata isn't deterministic between runs, but we shall see if that's the case.

**Comment by niklasdewally** [comment]
> > I think metadata is internal, so we should not hold ourselves to specific values or syntax for this in the integration tests. We also probably won't have the same metadata at the "rewrite check" stage every time,
> 
> I say keep it for now and update the expected files. Following the "err on the side of testing too much" rule.
> 
> I'll be surprised if metadata isn't deterministic between runs, but we shall see if that's the case.

👍🏻 

https://serde.rs/field-attrs.html

it seems to just be `#[serde(skip)]` to not serialise it, so it wouldn't be hard to do either way.

**Comment by ozgurakgun** [comment]
@Kieranoski702 is this blocked by anything? I think what's in this PR looks good!

**Comment by Kieranoski702** [comment]
This PR should be ready to be merged now

**Pull request merged**